### PR TITLE
NXP-29985: add the document deletion feature

### PR DESCRIPTION
**Pull request by RSalem07**

### CODE REVIEW

File: nuxeo-coldstorage-web/elements/nuxeo-delete-content-from-coldstorage-button.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
+/**
+ (C) Copyright Nuxeo Corp. (http://nuxeo.com/)
+
+ Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (efge) [inline nuxeo-coldstorage-web/elements/nuxeo-delete-content-from-coldstorage-button.js:4]: Is this a public or private code? For public code that's not the Copyright first line we use. For private code it's not under the Apache License.

File: nuxeo-coldstorage-web/i18n/messages.json [hunk 2]
```diff
@@ -15,5 +15,11 @@
   "moveDocumentsContentsToColdStorage.confirm": "Confirm",
   "moveDocumentsContentsToColdStorage.error": "Error moving the documents contents to cold storage.",
   "moveDocumentsContentsToColdStorage.success": "Documents contents moved to cold storage successfully.",
-  "moveDocumentsContentsToColdStorage.tooltip": "Send the selected items to cold storage"
-}
```
Comment (efge) [inline nuxeo-coldstorage-web/i18n/messages.json:19]: The closing `}` mustn't be removed.

File: nuxeo-coldstorage-web/i18n/messages.json [hunk 3]
```diff
@@ -15,5 +15,11 @@
...
+ "moveDocumentsContentsToColdStorage.tooltip": "Send the selected items to cold storage",
+
```
Comment (efge) [inline nuxeo-coldstorage-web/i18n/messages.json:19]: Indent properly + remove extra line

File: nuxeo-coldstorage-web/i18n/messages.json [hunk 4]
```diff
@@ -15,5 +15,11 @@
...
+  "deleteContentFromColdStorage.description" : "This will delete definitely the document content from cold storage.",
```
Comment (efge) [inline nuxeo-coldstorage-web/i18n/messages.json:20]: `definitely` is not proper English for the meaning you want, please have a native English speaker review the messages.

File: nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/helpers/ColdStorageHelper.java [hunk 5]
```diff
@@ -220,6 +222,19 @@ public static ColdStorageContentStatus checkColdStorageContentAvailability(CoreS
         return new ColdStorageContentStatus(beingRetrieved, available);
     }
 
+    /**
+     * Retrieves the {@link DocumentModel} of the given {@code id}.
+     *
+     * @param session the core session
+     * @param id
+    the document id
```
Comment (efge) [inline nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/helpers/ColdStorageHelper.java:230]: Put this on a single line.

File: nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/helpers/ColdStorageHelper.java [hunk 6]
```diff
@@ -220,6 +222,19 @@ public static ColdStorageContentStatus checkColdStorageContentAvailability(CoreS
...
+     * @return an optional containing the document model of the given {@code id} if it exists, empty otherwise
+     */
+
+    public static Optional<DocumentModel> getDocumentModelById(CoreSession session, DocumentRef documentIdRef) {
+        return Optional.of(session.getDocument(documentIdRef));
```
Comment (efge) [inline nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/helpers/ColdStorageHelper.java:235]: Here you would want `Optional.ofNullable` if you expect to be able to return an empty `Optional`.
But even this wouldn't work because if the document doesn't exist `session.getDocument` will not return `null` but throw a `DocumentNotFoundException`.
... which means that this code path hasn't been tested.
Finally I don't see the point of this one-line method used only once (which isn't even related to cold storage itself). You should just inline everything in the caller.

File: nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/helpers/ColdStorageHelper.java [hunk 7]
```diff
@@ -252,7 +267,7 @@ public int getTotalAvailable() {
         }
     }
 
-    private ColdStorageHelper() {
+    public ColdStorageHelper() {
         // no instance allowed
```
Comment (efge) [inline nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/helpers/ColdStorageHelper.java:271]: The point of these kinds of helpers is to not be instantiated, so this should stay `private`, as the comment says.

File: nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/DeleteFromcoldstorage.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
+package org.nuxeo.coldstorage.operations;
```
Comment (efge) [inline nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/DeleteFromcoldstorage.java:1]: Missing copyright header.

File: nuxeo-coldstorage-web/nuxeo-coldstorage.html [hunk 9]
```diff
@@ -20,10 +20,19 @@
   </template>
 </nuxeo-slot-content>
 
+<nuxeo-slot-content name="moveToColdStorageDocumentContentAction" slot="DOCUMENT_ACTIONS" order="121">
+  <template>
+    <nuxeo-delete-contents-from-coldstorage-button document="[[document]]"></nuxeo-delete-contents-from-coldstorage-button
+  </template>
+</nuxeo-slot-content>
+
+
```
Comment (efge) [inline nuxeo-coldstorage-web/nuxeo-coldstorage.html:29]: I can't really judge Web UI code, but I don't think there's a need for two empty lines here.

File: nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/DeleteFromcoldstorage.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+import org.nuxeo.ecm.automation.core.annotations.Operation;
+import org.nuxeo.ecm.automation.core.annotations.OperationMethod;
+import org.nuxeo.ecm.core.api.CoreSession;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.NuxeoException;
+import org.nuxeo.ecm.core.api.security.SecurityConstants;
+
+import java.util.Optional;
+
+import static javax.servlet.http.HttpServletResponse.SC_INTERNAL_SERVER_ERROR;
```
Comment (efge) [inline nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/DeleteFromcoldstorage.java:17]: Formatting import order is static first, then `java`, then the rest.

File: nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/DeleteFromcoldstorage.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+/**
+ * Delete the main content associated with the input {@link DocumentModel} from cold storage.
```
Comment (efge) [inline nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/DeleteFromcoldstorage.java:20]: `Delete` -> `Deletes`
"from cold storage" is isolated at the end so it's not clear if it relates to "the main content" or "the input DocumentModel".
-> Rephrase, for instance: `Deletes from cold storage the main content associated with the input {@link DocumentModel}`

File: nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/DeleteFromcoldstorage.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+ *
+ * @since 11.0
+ */
+@Operation(id = MoveToColdStorage.ID, category = Constants.CAT_BLOB, label = "Delete from Cold Storage", description = "Delete the main document content from the cold storage.")
+public class DeleteFromcoldstorage {
+
+    private static final Logger l = LogManager.getLogger(DeleteFromcoldstorage.ID);
```
Comment (efge) [inline nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/DeleteFromcoldstorage.java:27]: Our convention everywhere in the Nuxeo code is to call this `log`, and to have `getLogger(DeleteFromcoldstorage.class)`, not some id.

File: nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/DeleteFromcoldstorage.java [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+    public static final String ID = "Document.DeleteFromColdStorage";
+
+    @Context
+    protected CoreSession session;
+
+    @OperationMethod
+    public void run(DocumentModel d) {
```
Comment (efge) [inline nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/DeleteFromcoldstorage.java:35]: Nuxeo code uses `doc` (or sometimes `document` or `docModel`) but not `d`.

File: nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/DeleteFromcoldstorage.java [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+            throw new NuxeoException(String.format("The document: %s cannot be moved to cold storage"),
+
+
+
+                    SC_INTERNAL_SERVER_ERROR);
+
+        }
+
+        Optional<DocumentModel> o = ColdStorageHelper.getDocumentModelById(session, d.getRef());
+        DocumentModel dc = o.get();
```
Comment (efge) [inline nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/DeleteFromcoldstorage.java:48]: Why are you re-fetching the document that you already have?

File: nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/DeleteFromcoldstorage.java [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+        session.removeChildren(dc.getRef());
+        return;
```
Comment (efge) [inline nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/DeleteFromcoldstorage.java:50]: Remove redundant `return`;
But actually most of our operations that naturally act on a document also return it, so that it can be chained.

File: nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/DeleteFromcoldstorage.java [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+    private static final Logger l = LogManager.getLogger(DeleteFromcoldstorage.ID);
+
+    public static final String ID = "Document.DeleteFromColdStorage";
+
+    @Context
+    protected CoreSession session;
+
+    @OperationMethod
+    public void run(DocumentModel d) {
+        if (!session.hasPermission(d.getRef(), SecurityConstants.REMOVE)) {
```
Comment (efge) [inline nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/DeleteFromcoldstorage.java:36]: Why a manual security check? `session.removeChildren` already does its own security checks. If this security check does something more, there should be a comment explaining why.

File: nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/DeleteFromcoldstorage.java [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+            l.debug("The user {} does not have the right permissions to delete the content of document {}",
+                    session::getPrincipal, () -> d);
+            throw new NuxeoException(String.format("The document: %s cannot be moved to cold storage"),
+
+
+
+                    SC_INTERNAL_SERVER_ERROR);
+
```
Comment (efge) [inline nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/DeleteFromcoldstorage.java:44]: Why all these empty lines?

File: nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/DeleteFromcoldstorage.java [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+        DocumentModel dc = o.get();
+        session.removeChildren(dc.getRef());
+        return;
+    }
+
+}
+
+
+
+
```
Comment (efge) [inline nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/DeleteFromcoldstorage.java:57]: Remove useless empty lines.

File: nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/DeleteFromcoldstorage.java [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+                    SC_INTERNAL_SERVER_ERROR);
+
+        }
+
+        Optional<DocumentModel> o = ColdStorageHelper.getDocumentModelById(session, d.getRef());
+        DocumentModel dc = o.get();
+        session.removeChildren(dc.getRef());
```
Comment (efge) [inline nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/DeleteFromcoldstorage.java:49]: Removing children documents has nothing to do with removing the main content from cold storage.
I don't understand at all this operation.

File: nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/MoveToColdStorage.java [hunk 20]
```diff
@@ -76,12 +76,13 @@ public DocumentModel run(DocumentModel document) {
 
     @OperationMethod
     public DocumentModelList run(DocumentModelList documents) {
-        DocumentModelList result = new DocumentModelListImpl();
+        DocumentModelList result = null;
```
Comment (efge) [inline nuxeo-coldstorage/src/main/java/org/nuxeo/coldstorage/operations/MoveToColdStorage.java:79]: The changes in this file are unrelated to the ticket being fixed, please do a separate ticket/commit.

**Review by efge — CHANGES_REQUESTED** [review]

**Comment by efge** [comment]
There must be unit test for such a feature. Given the code of the operation, I don't understand at all what/why it's trying to do.

**Comment by efge** [comment]
Also the commit message is wrong, "add feature and commit" is not descriptive.

**Comment by RSalem07** [comment]
I Close the PR, sorry we didn't mention it, but this PR was created for a specific.
